### ISSUE TITLE
[Bug Report] SocketAsyncEventArgs will not be collected by GC. Memory Surge

### ISSUE THREAD

**Issue by RickyYCheng** [dotnet/runtime#92583]
### Description

Environment: Windows 11
SDKs: net6.0, net48 (due to the needs of work, I use netframework)

When using `TaskCompletionSource` in **while** loops and **System.Timers.Timer** class, the memory increase a lot.
To debug in the Visual Studio, we shall see a lot of objs were created but cannot collected by GC. 

Here is my async func using TaskCompletionSource: 
```cs
Task SendAsync(Socket socket, string msg)
{
    var tcs = new TaskCompletionSource<object>();
    var args = new SocketAsyncEventArgs { };
    var buffer = Encoding.UTF8.GetBytes(msg);
    args.SetBuffer(buffer, 0, buffer.Length);
    args.Completed += (s, e) =>
    {
        tcs.SetResult(true);
    };
    socket.SendAsync(args);
    return tcs.Task;
}
```

### Reproduction Steps

Create 2 projects (Console-app): one for test, the other for host.
# Project host: 
```cs
using System.Net;
using System.Net.Sockets;
using System.Text;

var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
listener.Bind(new IPEndPoint(IPAddress.Parse("127.0.0.1"), 8080));
listener.Listen(100);

var handler = await listener.AcceptAsync();
while (true)
{
    // Receive message.
    var buffer = new byte[1_024];
    var received = await handler.ReceiveAsync(buffer, SocketFlags.None);
    var response = Encoding.UTF8.GetString(buffer, 0, received);

    Console.WriteLine(response);
}
```
# project test
```cs
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;
using System.Timers;

var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
await socket.ConnectAsync("127.0.0.1", 8080);

//var timer = new Timer(1000) { AutoReset = true };
//timer.Elapsed += async (s, e) =>
//{
//    await SendAsync(socket, "Hello world!");
//};
//timer.Start();

while (true)
{
    await Task.Delay(1000);
    await SendAsync(socket, "Hello world!");
}

Task SendAsync(Socket socket, string msg)
{
    var tcs = new TaskCompletionSource<object>();
    var args = new SocketAsyncEventArgs { };
    var buffer = Encoding.UTF8.GetBytes(msg);
    args.SetBuffer(buffer, 0, buffer.Length);
    args.Completed += (s, e) =>
    {
        tcs.SetResult(true);
    };
    socket.SendAsync(args);
    return tcs.Task;
}
```
Steps: 
1. start project host 
2. start project test
3. take fast shot in VS to check memory and objects
4. switch the SDK version between net48 and net6.0
5. switch the code between timer and while loop

### Expected behavior

Memory should not increase so much, repeat `hello world!` each second.

### Actual behavior

For net6.0 while version: Log `Hello world!` only once
For other versions: memory increase a lot (unstopable)

Behaviors:
| ..........  |   net48   | net6.0 |
| While | increase | Behavior Error |
| Timer | increase | increase |

---

# net6.0 using timer 
```cs
var timer = new Timer(1000) { AutoReset = true };
timer.Elapsed += async (s, e) =>
{
    await SendAsync(socket, "Hello world!");
};
timer.Start();
```
![image](https://github.com/dotnet/runtime/assets/77863853/f4d54f98-f369-41e0-a601-e5b383cdb420)
![image](https://github.com/dotnet/runtime/assets/77863853/dd6b2efe-00c0-42c5-8208-7098ba07533a)
```txt
Hello world!
Hello world!
Hello world!
...
```
# net48 using while 
```cs
while (true)
{
    await Task.Delay(1000);
    await SendAsync(socket, "Hello world!");
}
```
![image](https://github.com/dotnet/runtime/assets/77863853/2f55e198-61e4-421d-b3af-3b11d07f2062)
![image](https://github.com/dotnet/runtime/assets/77863853/3ca87073-62ac-4a7b-89cb-4619cc7259fb)
```txt
Hello world!
Hello world!
Hello world!
...
```
# net48 using timer
![image](https://github.com/dotnet/runtime/assets/77863853/e15866e1-51ef-46c1-9346-f51766c19a54)
![image](https://github.com/dotnet/runtime/assets/77863853/f29af1f4-5b17-404e-a3a9-ea98fefcb2be)
```txt
Hello world!
Hello world!
Hello world!
...
```
# net 60 using while loops 
**Behavior Error**
```cs
while (true)
{
    await Task.Delay(1000);
    await SendAsync(socket, "Hello world!");
}
```
OUTPUT:
```txt
Hello world!
```
Should log multiple lines instead of only one. 

# net6.0 while loop without waiting 
```cs
while (true)
{
    await Task.Delay(1000);
    _ = SendAsync(socket, "Hello world!");
    // GC.Collect(); // for testing
}
```
![image](https://github.com/dotnet/runtime/assets/77863853/44ed0715-2428-4419-9689-f955022c2ecc)
> Memory ok, log message ok, but it is not awaitable

**Comment by to11mtm** [comment]
I think one potential issue could be that the `SocketAsyncEventArgs` is having an delegate added onto `Completed`, which likely is resulting in the TaskCompletionSource being held by said delegate, alongside the lack of clearing out of said delegate when completed. (Not sure whether `.Dispose()` would clear the handler immediately, doesn't look like it however)

It's *possible* that they are winding up somewhere that the GC isn't able to get to them easily. One heuristic way to check if this is the case, would be to add `GC.Collect(2)` calls into your test code. (But, this is not a fix suitable for production.)

Potential fixes to consider: 

```
    //option 1
    socket.SendAsync(args);
    //new code
    tcs.Task.ContinueWith(t=>{
        args.Completed = null;
    });
    //end of new code
    return tcs.Task;
```

```
   //Option 2
    args.Completed += (s, e) =>
    {
        tcs.SetResult(true);
        e.Completed = null;
    };
```

However both of these have drawbacks to consider such as allocations or possibly not firing (if `Completed` is never called) Typically `SocketAsyncEventArgs` are pooled or otherwise reused for max performance; otherwise the newer `Async` overloads are preferable since they handle some of that for you.

**Comment by RickyYCheng** [comment]
> See https://learn.microsoft.com/dotnet/api/system.net.sockets.socket.sendasync#system-net-sockets-socket-sendasync(system-net-sockets-socketasynceventargs)
> 
> > Returns
> > [Boolean](https://learn.microsoft.com/en-us/dotnet/api/system.boolean?view=net-7.0)
> > true if the I/O operation is pending. The [Completed](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socketasynceventargs.completed?view=net-7.0#system-net-sockets-socketasynceventargs-completed) event on the e parameter will be raised upon completion of the operation.
> > false if the I/O operation completed synchronously. In this case, The [Completed](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socketasynceventargs.completed?view=net-7.0#system-net-sockets-socketasynceventargs-completed) event on the e parameter will not be raised and the e object passed as a parameter may be examined immediately after the method call returns to retrieve the result of the operation.
> 
> You should check the return value of the `SendAsync` call, and signal the `tcs` if it returns false (as that means the callback will not be called as the operation is already done).
> 
> It is possible that the timings changed between versions, and now you are seeing more cases where the operation completes quickly, without having to invoke your callback, and your tasks are therefore never completing.

Thank you! I've found the problem, it not depends on the TaskCompletionSource. Acutally we can check 
```cs
args.Completed += (s, e) => 
{
  if (e.SocketError is SocketError.Success)
    ...
}
```
I will create a new issue to show. Plz chech the las comment.

**Comment by MihaZupan** [comment]
Are you calling `tcs.SetResult()` if `socket.SendAsync(args)` returned `false`?

**Comment by RickyYCheng** [comment]
To see the source code (decompiled by vs)
```cs
public void Dispose()
        {
            m_DisposeCalled = true;
            if (Interlocked.CompareExchange(ref m_Operating, 2, 0) == 0)
            {
                FreeOverlapped(checkForShutdown: false);
                GC.SuppressFinalize(this);
            }
        }

        ~SocketAsyncEventArgs()
        {
            FreeOverlapped(checkForShutdown: true);
        }
```
I noticed that the Finilizer and IDisposable.Dispose use the different arguments on `FreeOverlapped` func.
Since it is private, i am trying to use reflection to call it. Check:
```cs
public static class SocketAsyncEventArgsExtension
{
    static readonly Lazy<MethodInfo> finilize = 
        new Lazy<MethodInfo>(() => 
            typeof(SocketAsyncEventArgs).GetMethod("FreeOverlapped",
            BindingFlags.NonPublic | BindingFlags.Instance));
    public static void Finalize(this SocketAsyncEventArgs e)
        => finilize.Value?.Invoke(e, new object[] { true });
}
```
After calling the `SocketAsyncEventArgs` instance, 
```cs
async Task SendAsync(Socket socket, string msg)
{
    var tcs = new TaskCompletionSource<object>();
    var args = new SocketAsyncEventArgs { };
    var buffer = Encoding.UTF8.GetBytes(msg);
    args.SetBuffer(buffer, 0, buffer.Length);
    args.Completed += (s, e) =>
    {
        tcs.SetResult(e.SocketError is SocketError.Success);
    };
    socket.SendAsync(args);
    var result = await tcs.Task;
    args.Finalize();
}
```
> Attention last line.

Then we test again: 
```txt
Hello world!
Hello world!
Hello world!
Hello world!
```
> Console output
![image](https://github.com/dotnet/runtime/assets/77863853/8895f8a5-d449-4eb5-a9f5-98f22d391c93)
> memory and objects
---
So I think that the clause `FreeOverlapped(checkForShutdown: false);` in `IDisposable.Dispose` might be wrong. Hope to see more details.

**Comment by RickyYCheng** [comment]
> To see the source code (decompiled by vs)
> 
> ```cs
> public void Dispose()
>         {
>             m_DisposeCalled = true;
>             if (Interlocked.CompareExchange(ref m_Operating, 2, 0) == 0)
>             {
>                 FreeOverlapped(checkForShutdown: false);
>                 GC.SuppressFinalize(this);
>             }
>         }
> 
>         ~SocketAsyncEventArgs()
>         {
>             FreeOverlapped(checkForShutdown: true);
>         }
> ```
> 
> I noticed that the Finilizer and IDisposable.Dispose use the different arguments on `FreeOverlapped` func. Since it is private, i am trying to use reflection to call it. Check:
> 
> ```cs
> public static class SocketAsyncEventArgsExtension
> {
>     static readonly Lazy<MethodInfo> finilize = 
>         new Lazy<MethodInfo>(() => 
>             typeof(SocketAsyncEventArgs).GetMethod("FreeOverlapped",
>             BindingFlags.NonPublic | BindingFlags.Instance));
>     public static void Finalize(this SocketAsyncEventArgs e)
>         => finilize.Value?.Invoke(e, new object[] { true });
> }
> ```
> 
> After calling the `SocketAsyncEventArgs` instance,
> 
> ```cs
> async Task SendAsync(Socket socket, string msg)
> {
>     var tcs = new TaskCompletionSource<object>();
>     var args = new SocketAsyncEventArgs { };
>     var buffer = Encoding.UTF8.GetBytes(msg);
>     args.SetBuffer(buffer, 0, buffer.Length);
>     args.Completed += (s, e) =>
>     {
>         tcs.SetResult(e.SocketError is SocketError.Success);
>     };
>     socket.SendAsync(args);
>     var result = await tcs.Task;
>     args.Finalize();
> }
> ```
> 
> > Attentio the last line.
> 
> Then we test again:
> 
> ```
> Hello world!
> Hello world!
> Hello world!
> Hello world!
> ```
> 
> > Console output
> 
> ![image](https://user-images.githubusercontent.com/77863853/270621886-8895f8a5-d449-4eb5-a9f5-98f22d391c93.png)
> 
> > memory and objects
> 
> So I think that the clause `FreeOverlapped(checkForShutdown: false);` in `IDisposable.Dispose` might be wrong. Hope to see more details.

**Comment by RickyYCheng** [comment]
> Are you calling `tcs.SetResult()` manually (outside of the `Completed` callback) if `socket.SendAsync(args)` returned `false`?

I will use `tcs.SetResult(e.SocketError is SocketError.Success);` and do not care about the return value of `SendAsync(args)` because it works better. (The return value presents that whether the procedure is blocking or not)
